# Physalia adaptation course

This repository includes scripts and data associated with the practical sessions of the Physalia Course about Adaptation Genomics
Prepared by Anna Tigano, Yann Dorant and Claire MÃ©rot for the course in May 2021

Excpet for day 1 in which you won't have the bam files, all the tutorial can be completed with the compressed files stored in the github page for each day (day 2 to day 5


## Logging on the AWS server from your computer

[AWS for Mac OS X and Linux users](https://github.com/clairemerot/physalia_adaptation_course/blob/2021/AWS_mac.md)


[AWS for windows users](https://github.com/clairemerot/physalia_adaptation_course/blob/2021/AWS_windows.md)


## Day1: Handling NGS data: From raw reads to SNPs matrix

Capelin: Data taken from Cayuela et al,2020. Molecular Ecology https://doi.org/10.1111/mec.15499
 
Genome assembly: For this course, we made a dummy assembly of about 90 MB (instead of about 500 MB) and 5 chromosomes (instead of 24).

Raw reads: Obtained by GBS methods, sequenced with IonTorrent.
 
## Day2: Population structure and confounding factors

## Day3: Outlier detection - Environmental associations

Data: We focus on 12 population from Canada for which there is almost no geographic structure but environmental variability

3-1: Genetic structure and LD-pruned data

3-2: Outlier of differentiation with two methods (Outflank & Baypass)

3-3: Genotype-Environnement Associations with two methods (Baypass & redundancy analysis)

## Day4: Accounting for Structural Variants

4-1: Investigating haplotypes blocks ( ~inversions?)

This tutorial include working on local PCA, but also calculation of LD, Fst and observed fraction of heterozygotes which may be useful in other contexts

4-2: Filtering duplicated loci in RAD-seq data ( ~ Copy number variants)

## Day5: Functional approaches

